src/H5FD.c

```diff
@@ -1142,6 +1144,146 @@ H5FDfree(H5FD_t *file, H5FD_mem_t type, hid_t dxpl_id, haddr_t addr, hsize_t siz
     FUNC_LEAVE_API(ret_value)
 } /* end H5FDfree() */
 
+/*----------------------------------------------------------------------------
+ * Function:    H5FD__error_matches
+ *
+ * Purpose:     Checks if the most recent frame of the error stack have a
+ *              major and minor numbers that match the provided arguments.
+ *
+ * Return:      TRUE if arguments match the most recent error, FALSE otherwise
+ *
+ *----------------------------------------------------------------------------
+ */
+static hbool_t
+H5FD__error_matches(hid_t major, hid_t minor)
+{
+    H5E_t *estack = H5E__get_my_stack();
+    hid_t maj_type = estack->slot[estack->nused-1].maj_num;
+    hid_t min_type = estack->slot[estack->nused-1].min_num;
+    return maj_type == major && min_type == minor;
+}
+
+/*-------------------------------------------------------------------------
+ * Function:    H5FDalloc_mem
+ *
+ * Purpose:     Allocates memory for I/O. In the absence of a driver
+ *              callback implementation, system memory is used. Otherwise
+ *              the driver may resort to the use of a special memory
+ *              allocator that best suits the underlying device. Memory
+ *              allocated by this callback must be freed by a call to
+ *              H5FDfree_mem().
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5FDalloc_mem(H5FD_t *file, H5FD_ctl_alloc_args_t *args, void **out)
+{
+    herr_t ret_value = SUCCEED; /* Return value */
+    hbool_t use_system_memory = FALSE;
+
+    FUNC_ENTER_API(FAIL)
+
+    /* Check arguments */
+    if (!file)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "file pointer cannot be NULL")
+    if (!file->cls)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "file class pointer cannot be NULL")
+    if (!args)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "args pointer cannot be NULL")
+    if (!out)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "pointer to output buffer cannot be NULL")
+
+    if (file->cls->ctl) {
+        /* Delegate memory allocation to the VFD driver */
+        uint64_t flags = H5FD_CTL__FAIL_IF_UNKNOWN_FLAG;
+        if ((file->cls->ctl)(file, H5FD_CTL__MEM_ALLOC, flags, args, out) < 0) {
```

Comment:
Rather than directly calling the ctl callback here, this should probably route through H5FD_ctl. Though, you probably will still need to check whether the driver has a ctl callback or not here so that you can properly allocate system memory in the case that the driver doesn't support this. Also, have you considered switching H5E_FCNTL to H5E_UNSUPPORTED from our discussion? That seems safer than checking H5E_FCNTL here.

Reply:
Thanks for your comments, @jhendersonHDF. Yes, I did consider switching to H5E_UNSUPPORTED but ended up not bringing that one to this PR. I'll bring it to the next one.